google/cloud/forseti/scanner/audit/retention_rules_engine.py

```diff
@@ -0,0 +1,391 @@
+            'Lack of minimum_retention and '
+            'maximum_retention in rule {}'.format(rule_index))
+    elif minimum_retention != None and maximum_retention != None:
+        if minimum_retention > maximum_retention:
+            raise audit_errors.InvalidRulesSchemaError(
+                'minimum_retention larger than '
+                'maximum_retention in rule {}'.format(rule_index))
+    return minimum_retention, maximum_retention
+
+
+class RetentionRuleBook(bre.BaseRuleBook):
+    """The RuleBook for Retention resources."""
+
+    def __init__(self, rule_defs=None):
+        """Initialization.
+
+        Args:
+            rule_defs (dict): rule definitons
+        """
+        super(RetentionRuleBook, self).__init__()
+        self._rules_sema = threading.BoundedSemaphore(value=1)
+
+        self.resource_rules_map = {
+            applies_to: collections.defaultdict(set)
+            for applies_to in SUPPORTED_RETENTION_RES_TYPES}
+        if not rule_defs:
+            self.rule_defs = {}
+        else:
+            self.rule_defs = rule_defs
+            self.add_rules(rule_defs)
+
+    def add_rules(self, rule_defs):
+        """Add rules to the rule book
+
+        Args:
+            rule_defs (dict): rule definitions dictionary
+        """
+        for (i, rule) in enumerate(rule_defs.get('rules', [])):
+            self.add_rule(rule, i)
+
+    def add_rule(self, rule_def, rule_index):
+        """Add a rule to the rule book.
+
+        Args:
+            rule_def (dict): A dictionary containing rule definition
+                properties.
+            rule_index (int): The index of the rule from the rule definitions.
+                Assigned automatically when the rule book is built.
+        """
+        self._rules_sema.acquire()
+        try:
+            applies_to = rule_def.get('applies_to')
+            if applies_to is None:
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Lack of applies_to in rule {}'.format(rule_index))
+
+            minimum_retention, maximum_retention = get_retention_range(
+                rule_def, rule_index)
+
+            if any(support_type not in SUPPORTED_RETENTION_RES_TYPES
+                   for support_type in applies_to):
+                raise audit_errors.InvalidRulesSchemaError(
+                    'Invalid applies_to resource in rule {}'.format(rule_index))
+
+            added_applies_to = set()
+
+            for appto in applies_to:
+                if appto in added_applies_to:
+                    raise audit_errors.InvalidRulesSchemaError(
+                        'redundant applies_to in rule {}'.format(rule_index))
+                added_applies_to.add(appto)
+
+                self.create_and_add_rule(
+                    rule_def,
+                    rule_index,
+                    appto,
+                    minimum_retention,
+                    maximum_retention)
+        finally:
```

Comment:
remove sem

Reply:
I believe the _rules_sema is not necessary but "audit_logging", "enabled_apis", "iam", "iap", and "log_sink" all use it in this way.